### PR TITLE
Move TailRec after Erasure.

### PR DESCRIPTION
**Pull request by sjrd**
This used not to be possible because it was too complicated for TailRec to understand trees produced by the pattern matcher. Now that patmat uses `Labeled` blocks instead of label-defs, it is trivial to do so: the argument of a `return` from a labeled block `lab` is in tail position if and only if the labeled block `lab` is
itself in tail position.

Running TailRec after erasure has two major benefits:

* it is much simpler, as it does not have to deal with type parameters, `TypeApply`s, and a bunch of other stuff.
* it supports polymorphic tail-recursive calls by construction (recursive calls whose receiver or method has different type parameters).

It also has one difficulty: it cannot see the *call-site* `@tailrec` annotations anymore. This is why we add a mini-phase to record such annotations as tree attachments.

Having TailRec after erasure will also be necessary to later make it use `Labeled` blocks and loops instead of label-defs itself. Indeed, in that case it will need to declare local `var`s for its term parameters, which would break path-dependent types within the method if it were done before erasure.

### CODE REVIEW

**Comment by smarter** [comment]
> It also has one difficulty: it cannot see the call-site @tailrec annotations anymore. This is why we add a mini-phase to record such annotations as tree attachments.

We try to avoid using tree attachments unless there's no other choice in Dotty since they're a bit too magic, and using them to communicate across phases seems especially scary. Can we try to find an alternative solution? I suggest simply changing erasure to not drop annotations and check if that breaks anything.

**Comment by sjrd** [comment]
Yeah, I'm not happy with the tree attachment either. I just wanted to make those tests pass rather than moving them to pending (because, you know, one doesn't just remove tests), but it certainly is open for suggestions.

I don't think I have the stomach to try and change erasure right now, though. I suggest we discuss that at the next meeting.

**Comment by sjrd** [comment]
After a discussion with @odersky today, I added a commit that simply removes the call-site `@tailrec` checks feature altogether. Needs discussion.

**Comment by nicolasstucki** [comment]
What is the incentive to remove the `@tailrec` at call site?

**Comment by sjrd** [comment]
Rebased.

**Comment by smarter** [comment]
test performance please

File: compiler/src/dotty/tools/dotc/transform/MixinOps.scala [hunk 1]
```diff
@@ -24,7 +24,7 @@ class MixinOps(cls: ClassSymbol, thisPhase: DenotTransformer)(implicit ctx: Cont
       name = member.name.stripScala2LocalSuffix,
       flags = member.flags &~ Deferred,
       info = cls.thisType.memberInfo(member)).enteredAfter(thisPhase).asTerm
-    res.addAnnotations(member.annotations)
+    res.addAnnotations(member.annotations.filter(_.symbol != defn.TailrecAnnot))
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/transform/MixinOps.scala]: Sorry, I missed this. Is this needed if you move `TailRec` after `Mixin` within the same group?

File: compiler/src/dotty/tools/dotc/transform/TailRec.scala [hunk 2]
```diff
@@ -195,84 +208,45 @@ class TailRec extends MiniPhase with FullParameterization {
...
+              if (this.method.owner.isClass) receiver :: arguments
+              else arguments
+
+            Apply(method, argumentsWithReceiver)
           }
-          else fail(defaultReason)
+          else fail("it is not in tail position")
         } else {
-          val receiverIsSuper = (method.name eq sym) && enclosingClass.appliedRef.widen <:< recvWiden
+          val receiverIsSuper = (method.name eq sym) && enclosingClass.appliedRef.widen <:< prefix.tpe.widenDealias
```
[the remark this answers is not in the document]
Comment (allanrenucci) [reply compiler/src/dotty/tools/dotc/transform/TailRec.scala]: Sure! Do you mind opening an issue, once the PR is merged?

File: compiler/src/dotty/tools/dotc/transform/MixinOps.scala (hunk 1 above)
Comment (sjrd) [reply compiler/src/dotty/tools/dotc/transform/MixinOps.scala]: Even more so. But right now it's necessary above all for *extension methods* (for `AnyVal`s, which also use this `forwarder()` method). In general, when creating a forwarder of any sort, we must remove the `@tailrec` annotation.

Previously, no forwarder was ever created before `TailRec` ran. But now that `TailRec` runs after the creation of extension methods, we must ensure that `forwarder` drops `@tailrec`).

File: compiler/src/dotty/tools/dotc/transform/TailRec.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (sjrd) [reply compiler/src/dotty/tools/dotc/transform/TailRec.scala]: #5163

**Pull request merged**